Fix #5086 : Implement Dark Mode support for NPS Survey

MohitGupta121 opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation

Fix #5086 : Implement full Dark Mode support for NPS Survey

<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

### NPS Survey Welcome Screen

| Light Mode | Dark Mode |
| ------ | ------ |
| <img src="https://github.com/oppia/oppia-android/assets/76530270/0cad1b6a-6927-44ee-94e9-1fc6fd79d703" height="400" style="max-width: 100%">  | <img src="https://github.com/oppia/oppia-android/assets/76530270/8e5216f5-e8d0-4720-9187-4382d9aa3701" height="400" style="max-width: 100%"> |

### NPS Survey Question screen (Button Disabled)

| Light Mode | Dark Mode |
| ------ | ------ |
| <img src="https://github.com/oppia/oppia-android/assets/76530270/f03c3773-8a1e-484d-9630-7d1305c0a9d7" height="400" style="max-width: 100%">  | <img src="https://github.com/oppia/oppia-android/assets/76530270/4728e8a9-4de6-4db7-a022-64921a4179e3" height="400" style="max-width: 100%"> |

### NPS Survey Question screen (Button Enabled)

| Light Mode | Dark Mode |
| ------ | ------ |
| <img src="https://github.com/oppia/oppia-android/assets/76530270/c331cea5-d80c-404c-b4c9-a60540661136" height="400" style="max-width: 100%">  | <img src="https://github.com/oppia/oppia-android/assets/76530270/768d6d55-532b-4c6d-b939-2fd0df5667f9" height="400" style="max-width: 100%"> |

### NPS Survey Score Screen

| Light Mode | Dark Mode |
| ------ | ------ |
| <img src="https://github.com/oppia/oppia-android/assets/76530270/16af2177-db90-4370-a1e9-e04cec0677a0" height="400" style="max-width: 100%">  | <img src="https://github.com/oppia/oppia-android/assets/76530270/0cfeb8a6-6c52-4177-9b91-f6854a52d3b8" height="400" style="max-width: 100%"> |

### NPS Survey Final comment screen (Button Enabled)

| Light Mode | Dark Mode |
| ------ | ------ |
| <img src="https://github.com/oppia/oppia-android/assets/76530270/790b665f-a3ca-4b12-b04f-c60be6159632" height="400" style="max-width: 100%">  | <img src="https://github.com/oppia/oppia-android/assets/76530270/0007c758-3900-4a7e-a9ab-da536125f06a" height="400" style="max-width: 100%"> |

### NPS Survey Final comment screen (Button Disabled)

| Light Mode | Dark Mode |
| ------ | ------ |
| <img src="https://github.com/oppia/oppia-android/assets/76530270/a26e9da7-0b4b-4357-96ba-fee361daeba2" height="400" style="max-width: 100%">  | <img src="https://github.com/oppia/oppia-android/assets/76530270/08695850-f8fb-4227-80bd-c4f1cd35640f" height="400" style="max-width: 100%"> |



### NPS Survey Thanks You screen

| Light Mode | Dark Mode |
| ------ | ------ |
| <img src="https://github.com/oppia/oppia-android/assets/76530270/ec5d694e-798d-41d2-9ca7-aacfd23bf00e" height="400" style="max-width: 100%">  | <img src="https://github.com/oppia/oppia-android/assets/76530270/13fa3fc2-3f00-470e-87bf-7c1906a06677" height="400" style="max-width: 100%"> |

### NPS Survey Exit Dialog

| Light Mode | Dark Mode |
| ------ | ------ |
| <img src="https://github.com/oppia/oppia-android/assets/76530270/317385d3-5365-443c-905c-6a91457bbfa9" height="400" style="max-width: 100%">  | <img src="https://github.com/oppia/oppia-android/assets/76530270/2fbbed49-90b0-42e6-a579-d5aad9675057" height="400" style="max-width: 100%"> |



<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing
-

MohitGupta121 committed:
Implement dark mode for nps survey

MohitGupta121 reviewed (commented):
@adhiamboperes PTAL, Thanks!

adhiamboperes reviewed (changes requested):
Thanks a lot @MohitGupta121! I just had two minor comments:
- Can I see how the 'NPS Survey Final comment screen' looks without text and the submit button disabled?
- Is there a way to make the progress bar look like the one in light mode (I don't mind too much, but let's give it a try).

Otherwise this looks really great!

MohitGupta121 committed:
Merge branch 'develop' into nps_survey_dark_mode

MohitGupta121 commented:
@adhiamboperes PTAL Thanks, I updated the PR description with button disabled final comment screen.

2. Regarding progress bar I don't think there is need to change the theme/ color of that because we using the same progress bar in Exploration Fragments also. So I think that already suggested by designer in past. If we change then it's look in-consistence in both places. Actually now in question fragment we not showing progress bar so here is the figma design by designer for [Dark-Mode](https://xd.adobe.com/view/c05e9343-60f6-4c11-84ac-c756b75b940f-950d/screen/0d9126ed-6c55-4d47-b566-dc05a230a53c/) and [Light-Mode](https://xd.adobe.com/view/3dca36c2-5115-419c-b25e-0f10526b077c-6899/screen/4d4e4bd8-3755-41fc-bf83-3dcff5d5eee1/)


Also I think we need to introduce NPS Survey layout for landscape version because you can see the output, screen is not scrollable like welcome screen is incomplete and cutting from bottom:

![Screenshot_1690294206](https://github.com/oppia/oppia-android/assets/76530270/76842347-ef1e-440e-8197-14e749e9b2fb)
![Screenshot_1690294222](https://github.com/oppia/oppia-android/assets/76530270/81159558-3543-4c76-a645-acfb6ee89266)

adhiamboperes commented:
> @adhiamboperes PTAL Thanks, I updated the PR description with button disabled final comment screen.
> 
> 2. Regarding progress bar I don't think there is need to change the theme/ color of that because we using the same progress bar in Exploration Fragments also. So I think that already suggested by designer in past. If we change then it's look in-consistence in both places. Actually now in question fragment we not showing progress bar so here is the figma design by designer for [Dark-Mode](https://xd.adobe.com/view/c05e9343-60f6-4c11-84ac-c756b75b940f-950d/screen/0d9126ed-6c55-4d47-b566-dc05a230a53c/) and [Light-Mode](https://xd.adobe.com/view/3dca36c2-5115-419c-b25e-0f10526b077c-6899/screen/4d4e4bd8-3755-41fc-bf83-3dcff5d5eee1/)

I created a custom toolbar for the survey, please check the [layout file](https://github.com/oppia/oppia-android/blob/df820f7364b1383993f47fc3f9b320c3557d5835/app/src/main/res/layout/survey_fragment.xml#L27C7-L27C7). The progressbar [drawable](https://github.com/oppia/oppia-android/blob/df820f7364b1383993f47fc3f9b320c3557d5835/app/src/main/res/layout/survey_fragment.xml#L50) in this toolbar is also customized, per the survey mocks. See how it looks in the light mode screenshots with the white outline.

> Also I think we need to introduce NPS Survey layout for landscape version because you can see the output, screen is not scrollable like welcome screen is incomplete and cutting from bottom:

Yes, that is correct.

MohitGupta121 commented:
@adhiamboperes so now dark-mode for toolbar is look good to you? Now I have to just give you toolbar with outline in lightmode right?

MohitGupta121 committed:
progress bar survey dark-mode

MohitGupta121 committed:
Merge branch 'develop' into nps_survey_dark_mode

MohitGupta121 reviewed (commented):
@adhiamboperes PTAL, here are the screenshots of the progress bar in both light and dark-mode. Thanks!

| Light Mode | Dark Mode |
| ------ | ------ |
| <img src="https://github.com/oppia/oppia-android/assets/76530270/77c15282-df9d-43f2-9a49-d7fd898e937b" height="400" style="max-width: 100%">  | <img src="https://github.com/oppia/oppia-android/assets/76530270/41a35a0b-8e21-49c6-ab86-209099f6def0" height="400" style="max-width: 100%"> |

adhiamboperes commented:
> @adhiamboperes PTAL, here are the screenshots of the progress bar in both light and dark-mode. Thanks!
> 
> Light Mode	Dark Mode
> <img alt="" height="400" src="https://user-images.githubusercontent.com/76530270/256019813-77c15282-df9d-43f2-9a49-d7fd898e937b.png">	<img alt="" height="400" src="https://user-images.githubusercontent.com/76530270/256019808-41a35a0b-8e21-49c6-ab86-209099f6def0.png">

Thank you, this LGTM

MohitGupta121 commented:
Thank you so much @adhiamboperes anything else I need to update in this PR?

MohitGupta121 committed:
Merge branch 'develop' into nps_survey_dark_mode

adhiamboperes commented:
> Thank you so much @adhiamboperes anything else I need to update in this PR?

Hey @MohitGupta121, I have opened a PR for the landscape layouts [here](https://github.com/oppia/oppia-android/pull/5110) and assigned it to you and @kkmurerwa for review. If you are happy with it, we will merge it into this branch before merging to develop.

MohitGupta121 commented:
Okay sure @adhiamboperes I'll have a look into it.

adhiamboperes committed:
Fix Part of #5086: Add Layout Land for Survey UI (#5110)

## Explanation
Fix Part of #5086: We found that the navigation buttons in some
questions were not displaying correctly in landscape mode on phones.
This issue affects the survey popup as well. See
https://github.com/oppia/oppia-android/pull/5104#issuecomment-1649935209.

This PR introduces dedicated landscape layouts for the affected screens.

There are some minor bug fixes/ ux improvements also added that are
related to the survey behaviour:
1. Make the exit confirmation dialog non-cancelable so that the user
makes use of the dialog buttons.
2.Prevent accidentally exiting the survey by showing the exit
confirmation dialog on backpress.
3. Resume normal flow on exiting an exploration, when the survey is not
to be shown.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If
this PR fixes part of an issue, prefix the title with "Fix part of
#bugnum: ...".)
- [ ] Any changes to
[scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets)
files have their rationale included in the PR explanation.
- [x] The PR follows the [style
guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [ ] The PR does not contain any unnecessary code changes from Android
Studio
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and
is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

### Before 

![image](https://github.com/oppia/oppia-android/assets/59600948/46b76a8e-82f3-4dd3-9f95-d83268ccf9a5)

### After
![Screenshot 2023-07-28 at 00 25
19](https://github.com/oppia/oppia-android/assets/59600948/c9518e0b-d610-4532-91ce-d23b80e4dd1a)

### Before+Light Mode, After+Dark Mode
![Screenshot 2023-07-28 at 01 32
02](https://github.com/oppia/oppia-android/assets/59600948/89fbbdb5-7905-4afb-abb4-6bfd041466ae)

### Before+Light Mode, After+Dark Mode
![Screenshot 2023-07-28 at 01 40
50](https://github.com/oppia/oppia-android/assets/59600948/88ba0852-a95d-4b86-bc2f-2682da5908b0)

adhiamboperes commented:
Hi @MohitGupta121, in addition to the feedback email I shared, please update all the buttons including the ones in the dialogs to use sentence case. Thanks!

adhiamboperes commented:
> Also I think we need to introduce NPS Survey layout for landscape version because you can see the output, screen is not scrollable like welcome screen is incomplete and cutting from bottom:

I Updated the PR to fix this. Here are the screenshots
### Before 
![image](https://github.com/oppia/oppia-android/assets/59600948/46b76a8e-82f3-4dd3-9f95-d83268ccf9a5)

### After
![Screenshot 2023-07-28 at 00 25 19](https://github.com/oppia/oppia-android/assets/59600948/c9518e0b-d610-4532-91ce-d23b80e4dd1a)

### Before+Light Mode, After+Dark Mode
![Screenshot 2023-07-28 at 01 32 02](https://github.com/oppia/oppia-android/assets/59600948/89fbbdb5-7905-4afb-abb4-6bfd041466ae)

### Before+Light Mode, After+Dark Mode
![Screenshot 2023-07-28 at 01 40 50](https://github.com/oppia/oppia-android/assets/59600948/88ba0852-a95d-4b86-bc2f-2682da5908b0)

MohitGupta121 committed:
fix designer changes

MohitGupta121 committed:
restore minor changes

MohitGupta121 committed:
Merge branch 'develop' into nps_survey_dark_mode

MohitGupta121 commented:
@adhiamboperes PTAL, I make changes as per designer email. Thanks!

<img src="https://github.com/oppia/oppia-android/assets/76530270/125382b1-fb46-4949-85b6-095e9df1c418" height="400" style="max-width: 100%">  

<img src="https://github.com/oppia/oppia-android/assets/76530270/6cdcaf9c-2aa8-4fd3-b1e6-e8a947b7ee11" height="400" style="max-width: 100%">

<img src="https://github.com/oppia/oppia-android/assets/76530270/25633d17-8345-4df9-85cf-b5d0d7290795" height="400" style="max-width: 100%">

<img src="https://github.com/oppia/oppia-android/assets/76530270/63f24b25-5dc2-4934-8c56-2404164b1196" height="400" style="max-width: 100%">  

<img src="https://github.com/oppia/oppia-android/assets/76530270/8811c90c-8a7c-4292-9b44-5eee3c883ded" height="400" style="max-width: 100%">

adhiamboperes reviewed (approved):
Thanks @MohitGupta121, These changes LGTM!

adhiamboperes commented on app/src/main/res/drawable/nps_survey_button_background.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+  android:shape="rectangle">
+  <corners android:radius="@dimen/spotlight_done_button_corner_radius" />
```
Prefer creating a dedicated dimension here, instead of using one named after a different screen, per #2399.

MohitGupta121 replied:
Done.

adhiamboperes commented on app/src/main/res/drawable/nps_survey_button_inactive_background.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+  android:shape="rectangle">
+  <corners android:radius="@dimen/state_previous_next_button_radius" />
```
Check the dimensions file for a survey screen-specific radius, or create one.

MohitGupta121 replied:
Done.

adhiamboperes reviewed (commented):
Left a couple of minor follow up comments

MohitGupta121 committed:
make separate dimension for survey

MohitGupta121 committed:
Merge remote-tracking branch 'upstream/nps_survey_dark_mode' into nps_survey_dark_mode

MohitGupta121 commented:
@adhiamboperes done that separate dimensions changes for survey. Thanks!

seanlip commented:
@MohitGupta121 I think this looks great. Just one question: do we use dark green text on black anywhere else in the app (e.g. "Previous" on the black background)? I find that hard to read, but it's probably fine to leave as-is if it's an existing design pattern for dark mode (i.e. we've used it in other places too).

Thanks!

MohitGupta121 commented:
@seanlip PTAL, Thanks.
Here are some comparison:



| Current Green | Oppia Primary Green | Dark Green |
| ------ | ------ | ------ | 
| <img src="https://user-images.githubusercontent.com/76530270/257354462-6cdcaf9c-2aa8-4fd3-b1e6-e8a947b7ee11.png" height="400" style="max-width: 100%">  | <img src="https://github.com/oppia/oppia-android/assets/76530270/193a02a8-70a4-4c6b-9499-17fe1a9393e8" height="400" style="max-width: 100%"> | <img src="https://github.com/oppia/oppia-android/assets/76530270/8506caf5-d562-4ae0-9bdf-e6117805d38c" height="400" style="max-width: 100%"> |

seanlip commented:
@MohitGupta121 For me, the leftmost one is the most visible, but it is still a little dark against the background. 

What is the colour contrast ratio and does it meet accessibility standards? That's one way to test.

MohitGupta121 commented:
@seanlip Sorry, yes you are right. This does not match accessibility standards. Here is the output:
<img src="https://github.com/oppia/oppia-android/assets/76530270/5ef2a16e-b2ca-44dc-9002-a2f0e16edabb" height="400" style="max-width: 100%">

MohitGupta121 commented:
@seanlip After changing the color to oppia_bright_green here is the output screenshot and accessibility result:

### Screenshot:
<img src="https://github.com/oppia/oppia-android/assets/76530270/c18fe3bd-4dba-49f3-b6dd-4846067ef936" height="400" style="max-width: 100%"> 

### Accessibility Result
https://github.com/oppia/oppia-android/assets/76530270/00f56e00-00e1-4d0f-9114-102e838c01bc

MohitGupta121 committed:
fix night color of Previous button

seanlip reviewed (approved):
Thanks @MohitGupta121 ! No further concenrs.

adhiamboperes commented:
Thanks @MohitGupta121, @seanlip, @kaiyuxu.

I'm merging this now.
